File: robot_arm/实验1：正逆运动学/IK.m

function [wff] = IK(x,y,z,c,b,a)

d6=0.055;d1=0.284;a2=0.225;d4=0.2289;
T06=[cos(a)*cos(b) cos(a)*sin(b)*sin(c)-sin(a)*cos(c) cos(a)*sin(b)*cos(c)+sin(a)*sin(c) x
    sin(a)*cos(b) sin(a)*sin(b)*sin(c)+cos(a)*cos(c) sin(a)*sin(b)*cos(c)-cos(a)*sin(c) y
    -sin(b) cos(b)*sin(c) cos(b)*cos(c) z
    0 0 0 1];

Tt=T06;

trans=[1 0 0 0;0 0 -1 0;0 1 0 0; 0 0 0 1];
T06=T06*trans;

%calculate theta1  :2
w1(1)=atan((d6*T06(2,3)-y)/(d6*T06(1,3)-x));
w1(2)=w1(1)+pi;

%calculate theta3  :4
w3(1)=asin(((cos(w1(1))*x-d6*cos(w1(1))*T06(1,3)-T06(2,3)*sin(w1(1))*d6+sin(w1(1))*y)^2+(z-d1-d6*T06(3,3))^2-a2^2-d4^2)/(2*a2*d4));
w3(2)=pi-asin(((cos(w1(1))*x-d6*cos(w1(1))*T06(1,3)-T06(2,3)*sin(w1(1))*d6+sin(w1(1))*y)^2+(z-d1-d6*T06(3,3))^2-a2^2-d4^2)/(2*a2*d4));
w3(3)=asin(((cos(w1(2))*x-d6*cos(w1(2))*T06(1,3)-T06(2,3)*sin(w1(2))*d6+sin(w1(2))*y)^2+(z-d1-d6*T06(3,3))^2-a2^2-d4^2)/(2*a2*d4));
w3(4)=pi-asin(((cos(w1(2))*x-d6*cos(w1(2))*T06(1,3)-T06(2,3)*sin(w1(2))*d6+sin(w1(2))*y)^2+(z-d1-d6*T06(3,3))^2-a2^2-d4^2)/(2*a2*d4));

%calculate theta2  :8
w2=zeros(1,8);
for i=1:8
    t=mod(i,4);
    if t==0
        t=4;
    end
temp1=sqrt((d4*sin(w3(t))+a2)^2+d4*d4*cos(w3(t))*cos(w3(t)));
temp2=acos((d4*sin(w3(t))+a2)/temp1);

if mod(i,2)==1
w2(i)=asin((z-d6*T06(3,3)-d1)/temp1)+temp2;
else
   w2(i)=pi-(asin((z-d6*T06(3,3)-d1)/temp1)+temp2);
end

end

%calculate theta5  :16
w5=zeros(1,16);
for i=1:16
   t1=ceil(i/8);
   t3=ceil(i/4);
   t2=ceil(i/2);
if mod(i,2)==1
w5(i)=acos(T06(2,3)*sin(w1(t1))*sin(w2(t2)+w3(t3))+cos(w1(t1))*T06(1,3)*sin(w2(t2)+w3(t3))-T06(3,3)*cos(w2(t2)+w3(t3)));
else
   w5(i)=-acos(T06(2,3)*sin(w1(t1))*sin(w2(t2)+w3(t3))+cos(w1(t1))*T06(1,3)*sin(w2(t2)+w3(t3))-T06(3,3)*cos(w2(t2)+w3(t3)));
end
end


%calculate theta4 and theta 6 together  :32
w4=zeros(1,32);
w6=zeros(1,32);
w=zeros(32,6);
for i=1:32
    t1=ceil(i/16);
    t2=ceil(i/8);
    t3=ceil(i/4);
    t4=ceil(i/2);
if mod(i,2)==1
w4(i)=asin((-cos(w1(t1))*T06(2,3)+T06(1,3)*sin(w1(t1)))/sin(w5(t4)));
temp1=T06(1,2)*cos(w1(t1))*sin(w2(t3)+w3(t2))+T06(2,2)*sin(w1(t1))*sin(w2(t3)+w3(t2))-T06(3,2)*cos(w2(t3)+w3(t2));
w6(i)=asin(temp1/sin(w5(t4)));

else
w4(i)=pi-asin((-cos(w1(t1))*T06(2,3)+T06(1,3)*sin(w1(t1)))/sin(w5(t4)));
temp1=T06(1,2)*cos(w1(t1))*sin(w2(t3)+w3(t2))+T06(2,2)*sin(w1(t1))*sin(w2(t3)+w3(t2))-T06(3,2)*cos(w2(t3)+w3(t2));
w6(i)=pi-asin(temp1/sin(w5(t4)));

end
%put results into w
w(i,:)=[w1(t1) w2(t3)-pi/2 w3(t2) w4(i) w5(t4)+pi/2 w6(i)];
end


%delete the results out of scale
count=0;
wf=zeros(32,6);
for i=1:32

    if check(w(i,1),-pi,pi)==-1
        continue
    else
        w(i,1)=w(i,1)+pi*check(w(i,1),-pi,pi);
    end
    
    if check(w(i,2),-deg2rad(115),deg2rad(115))==-1
        continue
    else
        w(i,2)=w(i,2)+pi*check(w(i,2),-deg2rad(115),deg2rad(115));
    end
    
    if check(w(i,3),-deg2rad(130),deg2rad(130))==-1
        continue
    else
        w(i,3)=w(i,3)+pi*check(w(i,3),-deg2rad(40),deg2rad(220));
    end
    
    if check(w(i,4),-deg2rad(180),deg2rad(180))==-1
        continue
    else
        w(i,4)=w(i,4)+pi*check(w(i,4),-deg2rad(180),deg2rad(180));
    end
    
%     if check(w(i,5),-deg2rad(75),deg2rad(255))==-1
%         continue
%     else
%         w(i,5)=w(i,5)+pi*check(w(i,5),-deg2rad(75),deg2rad(255));
%     end
    if check(w(i,5),-0.78,3.92)==-1
        continue
    else
        w(i,5)=w(i,5)+pi*check(w(i,5),-0.78,3.92);
    end



    if check(w(i,6),-pi,pi)==-1
        continue
    else
        w(i,6)=w(i,6)+pi*check(w(i,6),-pi,pi);
    end
    count=count+1;
    wf(count,:)=w(i,:);
end
wf=wf(1:count,:);

%delete the results that could not be calculated by zhengyundongxue
Ncount=0;
wff=zeros(count,6);
for i=1:count
temp=DK(wf(i,:))-Tt;
t=norm(temp,1);
if t<1e-4
    Ncount=Ncount+1;
    wff(Ncount,:)=wf(i,:);
end

end
wff=wff(1:Ncount,:);
end

%check the rad and + -2pi with the scale
function [c]=check(w,a,b)
c=-1;
if (a<w)&&(w<b)
    c=0;
end

if (a<w+2*pi)&&(w+2*pi<b)
    c=2;
end

if (a<w-2*pi)&&(w-2*pi<b)
        c=-2;
end

end

File: robot_arm/probot_gazebo/src/IK.cpp
#include <iostream>
#include <cmath>
#include <Eigen/Eigen>
#include <Eigen/Dense>

using namespace std;
using namespace Eigen;
// #define pi 3.145926535897932384626

double pi = 2*acos(0);

class DK{
    public:
   Eigen::MatrixXd calDK();
   Eigen::MatrixXd calT(double alpha, double a, double d, double w);
   double w[6]{0};

    private:

};

Eigen::MatrixXd DK::calT(double alpha, double a, double d, double w){
    Eigen::MatrixXd T = Eigen::MatrixXd::Ones(4,4);

    T(0,0) = std::cos(w);
    T(0,1) = -std::sin(w);
    T(0,2) = 0;
    T(0,3) = a;
    T(1,0) = std::sin(w)*std::cos(alpha);
    T(1,1) = std::cos(w)*std::cos(alpha);
    T(1,2) = -std::sin(alpha);
    T(1,3) = -std::sin(alpha)*d;
    T(2,0) = std::sin(w)*std::sin(alpha);
    T(2,1) = std::cos(w)*std::sin(alpha);
    T(2,2) = std::cos(alpha);
    T(2,3) = cos(alpha)*d;
    T(3,0) = 0;
    T(3,1) = 0;
    T(3,2) = 0;
    T(3,3) = 1;

    return T;

}

Eigen::MatrixXd DK::calDK()
{
    Eigen::MatrixXd T = Eigen::MatrixXd::Identity(4,4);

    T =  T * calT(0,0,0.284,this->w[0]);
    T = T * calT(pi/2,0,0,pi/2+this->w[1]);
    T = T * calT(0,0.225,0,this->w[2]);
    T = T * calT(pi/2,0,0.2289,this->w[3]);
    T = T * calT(-pi/2,0,0,-pi/2+this->w[4]);
    T = T * calT(pi/2,0,0.055,this->w[5]);
    T = T * calT(-pi/2,0,0,0);

    return T;
}

class IK{
public:
    double pose[6]{0};//x,y,z,c,b,a
    double x = 0, y = 0, z = 0, c = 0, b = 0, a = 0;
    Eigen::Matrix4d T06;
    Eigen::MatrixXd Tt;
    double d6 = 0.055, d1 = 0.284, a2 = 0.225, d4 = 0.2289;
    void calT06();
    Eigen::MatrixXd cal_theta();
    int check_angular(double w, double a, double b);
 };

void IK::calT06()
{
    Eigen::MatrixXd T06(4,4);
    T06 << std::cos(a)*std::cos(b), std::cos(a)*std::sin(b)*std::sin(c)-std::sin(a)*std::cos(c), std::cos(a)*std::sin(b)*std::cos(c)+std::sin(a)*std::sin(c),  x,
    std::sin(a)*std::cos(b), std::sin(a)*std::sin(b)*std::sin(c)+std::cos(a)*std::cos(c), std::sin(a)*std::sin(b)*std::cos(c)-std::cos(a)*std::sin(c), y,
    -std::sin(b), std::cos(b)*std::sin(c), std::cos(b)*std::cos(c), z,
    0,0,0,1;

    this->Tt = T06;
    Eigen::MatrixXd trans(4,4);
    trans << 1,0,0,0,
    0,0,-1,0,
    0,1,0,0,
    0,0,0,1;

    this->T06 = T06 * trans;
    return;
}

Eigen::MatrixXd IK::cal_theta()
{
    //calculate theta 1
    double theta1[2]{0};
    theta1[0] = atan((d6*T06(1,2)-y) / (d6*T06(0,2)-x));
    theta1[1] = theta1[0] + pi;

    //calculate theta 3
    double theta3[4]{0};
    theta3[0] = asin((pow((std::cos(theta1[0])*x-d6*std::cos(theta1[0])*T06(0,2)-T06(1,2)*std::sin(theta1[0])*d6+std::sin(theta1[0])*y),2)+pow((z-d1-d6*T06(2,2)),2)-pow(a2,2)-pow(d4,2))/(2*a2*d4));
    theta3[1] = pi - asin((pow((std::cos(theta1[0])*x-d6*std::cos(theta1[0])*T06(0,2)-T06(1,2)*std::sin(theta1[0])*d6+std::sin(theta1[0])*y),2)+pow((z-d1-d6*T06(2,2)),2)-pow(a2,2)-pow(d4,2))/(2*a2*d4));
    theta3[2] = asin((pow((std::cos(theta1[1])*x-d6*std::cos(theta1[1])*T06(0,2)-T06(1,2)*std::sin(theta1[1])*d6+std::sin(theta1[1])*y),2)+pow((z-d1-d6*T06(2,2)),2)-pow(a2,2)-pow(d4,2))/(2*a2*d4));
    theta3[3] = pi -  asin((pow((std::cos(theta1[1])*x-d6*std::cos(theta1[1])*T06(0,2)-T06(1,2)*std::sin(theta1[1])*d6+std::sin(theta1[1])*y),2)+pow((z-d1-d6*T06(2,2)),2)-pow(a2,2)-pow(d4,2))/(2*a2*d4));

    //calculate theta2
    double theta2[8]{0};
    int t = 0;
    for(int i=0;i<8;i++)
    {
        t = i % 4;
        
        double temp1 = sqrt(pow((d4*std::sin(theta3[t])+a2),2)+d4*d4*std::cos(theta3[t])*cos(theta3[t]));
        double temp2 = acos((d4*sin(theta3[t])+a2)/temp1);

        if(i % 2 == 0)
        {
            theta2[i] = asin((z-d6*T06(2,2)-d1)/temp1)+temp2;
        }
        else
        {
            theta2[i] = pi - (asin((z-d6*T06(2,2)-d1)/temp1)+temp2);
        }

    }

    //calculate theta5
    double theta5[16]{0};
    for(int i=0;i<16;i++)
    {
        int t1 = i / 8;
        int t3 = i / 4;
        int t2 = i / 2;
        if(i % 2 == 0)
        {
            theta5[i] = acos(T06(1,2)*std::sin(theta1[t1])*std::sin(theta2[t2]+theta3[t3])+std::cos(theta1[t1])*T06(0,2)*std::sin(theta2[t2]+theta3[t3])-T06(2,2)*std::cos(theta2[t2]+theta3[t3]));
        }
        else{
            theta5[i] = -acos(T06(1,2)*std::sin(theta1[t1])*std::sin(theta2[t2]+theta3[t3])+std::cos(theta1[t1])*T06(0,2)*std::sin(theta2[t2]+theta3[t3])-T06(2,2)*std::cos(theta2[t2]+theta3[t3]));
        }
    }

    //calculate theta4 & theta6
    double theta4[32]{0};
    double theta6[32]{0};
    Eigen::MatrixXd theta(32,6);
    // double theta[32][6]{0};
    for(int i=0;i<32;i++)
    {
        int t1 = i / 16;
        int t2 = i / 8;
        int t3 = i /4;
        int t4 = i / 2;

        if(i % 2 == 0)
        {
            theta4[i] = asin((-std::cos(theta1[t1])*T06(1,2)+T06(0,2)*std::sin(theta1[t1]))/std::sin(theta5[t4]));
            double temp1 = T06(0,1)*std::cos(theta1[t1])*std::sin(theta2[t3]+theta3[t2])+T06(1,1)*std::sin(theta1[t1])*std::sin(theta2[t3]+theta3[t2])-T06(2,1)*std::cos(theta2[t3]+theta3[t2]);
            theta6[i] = asin(temp1/std::sin(theta5[t4]));
        }
        else{
            theta4[i] = pi - asin((-std::cos(theta1[t1])*T06(1,2)+T06(0,2)*std::sin(theta1[t1]))/std::sin(theta5[t4]));
            double temp1 = T06(0,1)*std::cos(theta1[t1])*std::sin(theta2[t3]+theta3[t2])+T06(1,1)*std::sin(theta1[t1])*std::sin(theta2[t3]+theta3[t2])-T06(2,1)*std::cos(theta2[t3]+theta3[t2]);
            theta6[i] = pi - asin(temp1/std::sin(theta5[t4]));
        }

        Eigen::MatrixXd temp(1,6);
        temp << theta1[t1], theta2[t3]-pi/2,theta3[t2], theta4[i], theta5[t4]+pi/2, theta6[i];

        theta.block<1,6>(i,0) = temp;
    }

    Eigen::MatrixXd result(32,6);
    int count = -1;
    for(int i=0;i<32;i++)
    {
        if (check_angular(theta(i,0), -pi, pi) == -1) 
        {
            continue;
        }
        else{
            theta(i,0) = theta(i,0) + check_angular(theta(i,0), -pi, pi) * pi;
        }
        if (check_angular(theta(i,1), -2.01, 2.01) == -1)
        {
            continue;
        }
        else{
            theta(i,1) = theta(i,1) + check_angular(theta(i,1), -2.01, 2.01) * pi;
        }
        if (check_angular(theta(i,2), -0.69, 3.83) == -1)
        {
            continue;
        }
        else{
            theta(i,2) = theta(i,2) + check_angular(theta(i,2), -0.69, 3.83) * pi;
        }
        if (check_angular(theta(i,3), -pi, pi) == -1) 
        {
            continue;
        }
        else{
            theta(i,3) = theta(i,3) + check_angular(theta(i,3), -pi, pi) * pi;
        }
        if (check_angular(theta(i,4), -0.78, 3.92) == -1)
        {
            continue;
        }
        else{
            theta(i,4) = theta(i,4) + check_angular(theta(i,4), -0.78, 3.92) * pi;
        }
        if (check_angular(theta(i,5), -pi, pi) == -1)
        {
            continue;
        }
        else{
            theta(i,5) = theta(i,5) + check_angular(theta(i,5), -pi, pi) * pi;
        }

        count += 1;
        Eigen::MatrixXd tmp(1,6);
        tmp << theta(i,0), theta(i,1), theta(i,2), theta(i,3), theta(i,4), theta(i,5);
        result.block<1,6>(count,0) =  tmp;

    }

    DK dk;
    Eigen::MatrixXd final_result(32,6);
    int cnt = 0;
    for(int i=0;i<=count;i++)
    {
        for(int j=0;j<6;j++)
        {
            dk.w[j] = result(i,j);
        }
        Eigen::MatrixXd T = dk.calDK();
        Eigen::MatrixXd dT = T - Tt;
        double sum = 0;
        for(int k=0;k<4;k++)
        {
            for(int m=0;m<4;m++)
            {
                sum += abs(dT(k,m));
            }
        }
        if (sum < 0.0001)
        {
            Eigen::MatrixXd tmp(1,6);
            tmp << result(i,0), result(i,1), result(i,2), result(i,3), result(i,4), result(i,5);
            final_result.block<1,6>(cnt,0) = tmp;
            cnt += 1;
            // std::cout<<cnt<<endl;
        }

    }

    Eigen::MatrixXd final_matrix = final_result.block(0, 0, cnt, 6);


    // Eigen::MatrixXd final_result(count+1,6);
    // final_result = result.block<count+1,6>(0,0);

    return final_matrix;
}

int IK::check_angular(double w, double a, double b)
{
    int flag = -1;
    if ((a < w) && (w < b))
    {
        flag = 0;
    }

    if ((a < w + 2 * pi) && (w + 2 * pi < b))
    {
        flag = 2;
    }

    if ((a < w - 2 * pi) && (w - 2 * pi < b))
    {
        flag = -2;
    }

    return flag;

}

int main(int argc, char **argv)
{
        std::cout<<"please input x,y,z,roll,pitch,yaw:"<<endl;
        IK ik;
        std::cin>>ik.x>>ik.y>>ik.z>>ik.c>>ik.b>>ik.a;
        ik.calT06();
        Eigen::MatrixXd theta;
        theta = ik.cal_theta();
        std::cout<<"Theta:"<<endl<<theta<<endl;
}
 